### ISSUE TITLE
bicep publish seems to be broken

### ISSUE THREAD

**Issue by GeraldGoldstein** [Azure/azure-cli#26241]
> ### `az feedback` auto-generates most of the information requested below, as of CLI version 2.0.62

**Related command**
variables:
  spn: SC-BSM-IAC-SharedServices-ContainerRegistry
  CarmlVersionNumber: 0.10.0
  RegistryName: ihubcarmlbicep.azurecr.io

      - task: AzureCLI@2
        displayName: Publish user assigned identity module.
        inputs:
          azureSubscription: $(spn)
          scriptType: 'bash'
          scriptLocation: 'inlineScript'
          inlineScript: |
            az bicep publish \
            --file modules/Microsoft.ManagedIdentity/userAssignedIdentities/deploy.bicep \
            --target br:$(RegistryName)/bicep/modules/microsoft.managedidentity/user_assigned_identities:$(CarmlVersionNumber) \

**Describe the bug**
ERROR: The command failed with an unexpected error. Here is the traceback:
ERROR: [Errno 2] No such file or directory: '/home/vsts/work/_temp/.azclitask/bin/bicep'
Traceback (most recent call last):
  File "/opt/az/lib/python3.10/site-packages/knack/cli.py", line 233, in invoke
    cmd_result = self.invocation.execute(args)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 663, in execute
    raise ex
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 726, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 697, in _run_job
    result = cmd_copy(params)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 333, in __call__
    return self.handler(*args, **kwargs)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/command_operation.py", line 121, in handler
    return op(**command_args)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/command_modules/resource/custom.py", line 3779, in publish_bicep_file
    if bicep_version_greater_than_or_equal_to(minimum_supported_version):
  File "/opt/az/lib/python3.10/site-packages/azure/cli/command_modules/resource/_bicep.py", line 202, in bicep_version_greater_than_or_equal_to
    installed_version = _get_bicep_installed_version(installation_path)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/command_modules/resource/_bicep.py", line 273, in _get_bicep_installed_version
    installed_version_output = _run_command(bicep_executable_path, ["--version"])
  File "/opt/az/lib/python3.10/site-packages/azure/cli/command_modules/resource/_bicep.py", line 311, in _run_command
    process = subprocess.run([rf"{bicep_installation_path}"] + args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
  File "/opt/az/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/opt/az/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/opt/az/lib/python3.10/subprocess.py", line 1847, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/home/vsts/work/_temp/.azclitask/bin/bicep'
To check existing issues, please visit: https://github.com/Azure/azure-cli/issues
To open a new issue, please run `az feedback`
##[error]Script failed with exit code: 1
/usr/bin/az account clear
Finishing: Publish user assigned identity module.

**To Reproduce**
running the publish command from ado agent throws this error.

**Expected behavior**
publish command should complete with a success message that the module was published to the registry.

**Environment summary**
take any bicep module, create a simple ADO pipeline, put in there an az cli task to publish the module and watch it fail.

**Additional context**

**Comment by yonzhan** [comment]
Thank you for opening this issue, we will look into it.

**Comment by GeraldGoldstein** [comment]
hi, just tried to do this on devcontainer and fails there as well, so this is not just ADO agent.

PS /workspaces/CARML> az bicep publish --file modules/Microsoft.ManagedIdentity/userAssignedIdentities/deploy.bicep --target "br:ihubcarmlbicep.azurecr.io/bicep/modules/microsoft.managedidentity/user_assigned_identities:0.10.0"
The command failed with an unexpected error. Here is the traceback:
[Errno 2] No such file or directory: '/home/vscode/.azure/bin/bicep'
Traceback (most recent call last):
  File "/opt/az/lib/python3.10/site-packages/knack/cli.py", line 233, in invoke
    cmd_result = self.invocation.execute(args)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 663, in execute
    raise ex
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 726, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 697, in _run_job
    result = cmd_copy(params)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/__init__.py", line 333, in __call__
    return self.handler(*args, **kwargs)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/core/commands/command_operation.py", line 121, in handler
    return op(**command_args)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/command_modules/resource/custom.py", line 3779, in publish_bicep_file
    if bicep_version_greater_than_or_equal_to(minimum_supported_version):
  File "/opt/az/lib/python3.10/site-packages/azure/cli/command_modules/resource/_bicep.py", line 202, in bicep_version_greater_than_or_equal_to
    installed_version = _get_bicep_installed_version(installation_path)
  File "/opt/az/lib/python3.10/site-packages/azure/cli/command_modules/resource/_bicep.py", line 273, in _get_bicep_installed_version
    installed_version_output = _run_command(bicep_executable_path, ["--version"])
  File "/opt/az/lib/python3.10/site-packages/azure/cli/command_modules/resource/_bicep.py", line 311, in _run_command
    process = subprocess.run([rf"{bicep_installation_path}"] + args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
  File "/opt/az/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/opt/az/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/opt/az/lib/python3.10/subprocess.py", line 1847, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/home/vscode/.azure/bin/bicep'
To check existing issues, please visit: https://github.com/Azure/azure-cli/issues
To open a new issue, please run `az feedback`
PS /workspaces/CARML>

**Comment by SwathiDhanwada-MSFT** [comment]
@GeraldGoldstein Thanks for your comment. We will review the issue and get back to you shortly.

**Comment by spoelstraethan** [comment]
I'm shocked this has been an issue this long, I just tried for most of the day to publish a module to a registry and have been getting the same error and I was sure it was my code and not a HUGE BUG in a CRITICAL component required for our IaC ONLY deployment policies..............

**Comment by spoelstraethan** [comment]
Ironically it appears by setting the `use_bicep_from_path` value to `true` instead of `false` ie `az config set bicep.use_binary_from_path=true
`, and then re-running my pipeline I was able to successfully publish again, and then even reinstalling/resetting the value to `false` with the full "fix" allowed me to work around this issue.

```
az bicep uninstall
az config set bicep.use_binary_from_path=false
az bicep install
az bicep publish --file VirtualMachine\update.bicep --target br:$(ourRegistry)/module/name:$(version)
```

**Comment by torreymicrosoft** [comment]
Thanks for your patience @GeraldGoldstein — this was recently picked up as an outstanding open issue for the ARM Deployments team to review. Apologies for the long wait on this one.

Re-reading the thread alongside the current `_bicep.py` and `custom.py` in the `dev` branch: it looks like the failure path you hit no longer exists. In 2023, `publish_bicep_file` jumped straight into `bicep_version_greater_than_or_equal_to()`, which tried to run the Bicep binary directly and raised `FileNotFoundError` whenever Bicep wasn't already installed in the path the AzureCLI@2 task was using. In the current source, `publish_bicep_file` now starts with an `ensure_bicep_installation(...)` call, so Bicep gets auto-installed (silently) before any version check or subprocess invocation can fire.

@spoelstraethan's 2023 workaround (`az bicep uninstall && az config set bicep.use_binary_from_path=false && az bicep install`) was mapping closely to what the code now does automatically — installing Bicep into the right place before `publish` ran. A specific CI-environment follow-up also landed in #31202 (merged April 2025).

If anyone is on a recent CLI (2.70+) and still seeing `FileNotFoundError` in `publish_bicep_file` on an ADO agent, please reopen with a fresh traceback — otherwise I've closed this out.
